In [78]:
from sklearn.model_selection import  RandomizedSearchCV, GridSearchCV,train_test_split
from sklearn import datasets
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import SGDClassifier

In [79]:
iris = datasets.load_iris()

In [80]:
df=pd.DataFrame(iris.data,columns=iris.feature_names)
df['target'] = iris.target

In [81]:
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [82]:
model_params_grid = {
    "MultinomialNB": {
        "classifier": MultinomialNB(),
        "param_grid": {
            "alpha": [0.001, 0.01, 0.1, 1.0],
        }
    },
    "LogisticRegression": {
        "classifier": LogisticRegression(multi_class="multinomial", max_iter=1000),
        "param_grid": {
            "tol": [0.001, 0.01, 0.1, 1.0],
            "C": [0.1, 1.0, 10.0],
            "solver": ["lbfgs"],
            "penalty": ["l2"], 
        }
    },
    "DecisionTreeClassifier": {
        "classifier": DecisionTreeClassifier(),
        "param_grid": {
            "max_depth": [1, 2, 3, 4],
        }
    },
    "RandomForestClassifier": {
        "classifier": RandomForestClassifier(),
        "param_grid": {
            "n_estimators": [100, 200],
            "max_depth": [1, 2, 3],
        }
    },
    "KNeighborsClassifier": {
        "classifier": KNeighborsClassifier(),
        "param_grid": {
            "n_neighbors": [1, 2, 3],
            "weights": ["uniform", "distance"],
        }
    },
    "SGDClassifier": {
        "classifier": SGDClassifier(random_state=42, shuffle=True),
        "param_grid": {
            "tol": [0.001, 0.01, 0.1, 1.0],
            "alpha": [0.0001, 0.001, 0.01],
            "penalty": ["l1", "l2"],
        }
    },
    "SVC": {
        "classifier": SVC(),
        "param_grid": {
            "tol": [0.001, 0.01, 0.1, 1.0],
            "C": [0.1, 1.0, 10.0],
        }
    }
}


In [83]:
x=df[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)']]
y=df['target']

In [84]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [85]:
reseult=[]
for item,items in model_params_grid.items():
    gs=GridSearchCV(items["classifier"],param_grid=items["param_grid"],cv=5,scoring="accuracy")
    gs.fit(x_train,y_train)
    accurcy=gs.score(x_test,y_test)
    reseult.append({
        "model":item,
        "best score":gs.best_score_,
        "best params":gs.best_params_,
    })
    

In [86]:
pd.DataFrame(reseult,columns=["model","best score","best params"])

model  best score  \
0           MultinomialNB    0.933333   
1      LogisticRegression    0.975000   
2  DecisionTreeClassifier    0.941667   
3  RandomForestClassifier    0.950000   
4    KNeighborsClassifier    0.958333   
5           SGDClassifier    0.941667   
6                     SVC    0.950000   

                                         best params  
0                                   {'alpha': 0.001}  
1  {'C': 10.0, 'penalty': 'l2', 'solver': 'lbfgs'...  
2                                   {'max_depth': 4}  
3              {'max_depth': 3, 'n_estimators': 100}  
4           {'n_neighbors': 3, 'weights': 'uniform'}  
5     {'alpha': 0.01, 'penalty': 'l1', 'tol': 0.001}  
6                           {'C': 1.0, 'tol': 0.001}

<h5> intercept </h5>
<h6>    the best model in logistic regression  with 0.975 accurcy score    </h6>

In [87]:
reseult=[]
for item , items  in model_params_grid.items() :
    rg=RandomizedSearchCV(items["classifier"],param_distributions=items["param_grid"],cv=5,scoring="accuracy",n_iter=2,random_state=42)
    rg.fit(x_train,y_train)
    accurcy=rg.score(x_test,y_test)
    reseult.append({
        "model":item,
        "best score":rg.best_score_,
        "best params":rg.best_params_,
    })


In [88]:
pd.DataFrame(reseult,columns=["model","best score","best params"])

model  best score  \
0           MultinomialNB    0.933333   
1      LogisticRegression    0.941667   
2  DecisionTreeClassifier    0.941667   
3  RandomForestClassifier    0.933333   
4    KNeighborsClassifier    0.950000   
5           SGDClassifier    0.941667   
6                     SVC    0.950000   

                                         best params  
0                                    {'alpha': 0.01}  
1  {'tol': 0.1, 'solver': 'lbfgs', 'penalty': 'l2...  
2                                   {'max_depth': 4}  
3              {'n_estimators': 200, 'max_depth': 1}  
4           {'weights': 'uniform', 'n_neighbors': 1}  
5     {'tol': 0.001, 'penalty': 'l1', 'alpha': 0.01}  
6                            {'tol': 0.1, 'C': 10.0}

<h5>  interception  </h5>
<h6>  i have two best model KNeighborsClassifier and Svc with 0.95 accurcy score </h6>